miden/tests/integration/stdlib/crypto/fri_all/verifier_fri_e2f4.rs

```diff
@@ -0,0 +1,459 @@
+use std::marker::PhantomData;
+use std::mem;
+
+use miden::math::fft;
+use miden::utils::math::log2;
+use miden::Digest as MidenDigest;
+use processor::Hasher;
+use prover::MerkleSet;
+use vm_core::chiplets::hasher::Hasher as MidenHasher;
+use vm_core::crypto::merkle::{MerklePath, MerklePathSet, NodeIndex};
+use vm_core::utils::{IntoBytes, RandomCoin};
+use vm_core::{Felt, FieldElement, QuadExtension, StarkField, ZERO};
+use winter_fri::folding::fold_positions;
+use winter_fri::{DefaultProverChannel, FriProver};
+use winter_fri::{FriOptions, FriProof, VerifierError};
+use winter_utils::group_vector_elements;
+
+use super::channel::MidenFriVerifierChannel;
+use super::channel::UnBatch;
+
+type QuadExt = QuadExtension<Felt>;
+
+pub fn fri_prove_verify_fold4_ext2(
+    trace_length_e: usize,
+) -> Result<
+    (
+        (Vec<MerkleSet>, Vec<([u8; 32], Vec<Felt>)>),
+        Vec<u64>,
+        Vec<u64>,
+        Vec<u64>,
+        Vec<u64>,
+        usize,
+    ),
+    VerifierError,
+> {
```

Comment:
Would be great to add some doc comments to this function. Also, the return type is rather complex - I'd probably create a wrapping struct with named fields to make it easier to read and use.

Reply:
Yes, makes sense. I add some more documentation at the beginning of this function. Will probably refactor it in a different PR later on if needed.